### DISCUSSION TITLE
Different codeowners required approvals if PR author is a codeowner or not

### DISCUSSION THREAD

**Discussion by ghost** [community/community#84831] [Pull Requests]
### Select Topic Area

Product Feedback

### Body

Let's say we have John from Team A and Julie from Team B. John and other his teammates from Team A are codeowners (CO) of repo Foo. I want to set a branch rule that requires 1 PR approval from CO if PR author is a CO (like John) and 2 CO approvals if author is not a CO (like Julie).

The motivation is that CO is a part of the team that daily maintains and develops the repository so his code, his decisions can be trusted enough to require less CO approvals. For users that are not part of codeowners we want a review from more CO.

**Comment by bonizario** [comment]
Unfortunately, GitHub does not currently support dynamic branch protection rules based on the author's status as a code owner. The branch protection rules are static and apply to all pull requests equally, regardless of who authored them.

However, you can achieve a similar effect using GitHub Actions. You can create a GitHub Action that runs when a pull request is opened. This action can check if the PR author is a code owner, and then add a label to the PR based on the result. You can then use these labels to communicate the required number of reviews.

Here's a rough idea of how you could set this up:

1. Create a GitHub Action that runs on the `pull_request` event.
2. In the action, use the GitHub API to get the list of code owners.
3. Check if the PR author is in the list of code owners.
4. If the author is a code owner, add a "1 approval required" label to the PR. If the author is not a code owner, add a "2 approvals required" label.

This won't enforce the approval requirements automatically, but it will make it clear to your team how many approvals are expected for each PR.

Remember, this is a workaround and not a perfect solution. The best way to handle this would be for GitHub to support dynamic branch protection rules. You might want to consider submitting a feature request to GitHub for this functionality.

**Reply by ghost** [reply]
Thank you for the idea. Labels are nice but if you work in a mature team. Unfortunately I don't. It's said but ma many developers just don't care and enforcing rules is the only way.

> consider submitting a feature request to GitHub for this functionality

How do I do that? Isn't Product Feedback a kind of feature request?

**Comment by bryan-shea** [comment]


**Reply by ghost** [reply]
Hi, thank you for detailed answer. About Step 5 "Status Checks". GA allows to create action like before_merge? It could be nice solution until GitHub implements my desired feature.

**Reply by bryan-shea** [reply]
Absolutely! GitHub Actions doesn't have a before_merge trigger, but setting up actions to run when a review is dropped on your PR could be a solid workaround. It's like having a gatekeeper right before the merge, making sure everything's in check. Pretty handy until GitHub rolls out something more specific for this! 👍🚀

**Reply by LandonTClipp** [reply]
Did you seriously just copy-paste a ChatGPT response? lol
